File: app/src/main/java/net/sytes/kai_soft/letsbuyka/Lists/ListsListFragment.java
package net.sytes.kai_soft.letsbuyka.Lists;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Canvas;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import net.sytes.kai_soft.letsbuyka.CRUDdb;
import net.sytes.kai_soft.letsbuyka.SwipeController;
import net.sytes.kai_soft.letsbuyka.Application;
import net.sytes.kai_soft.letsbuyka.Constants;
import net.sytes.kai_soft.letsbuyka.DataBase;
import net.sytes.kai_soft.letsbuyka.IFilterContract;
import net.sytes.kai_soft.letsbuyka.R;
import net.sytes.kai_soft.letsbuyka.SwipeControllerActions;

import java.util.ArrayList;

/**
 * Created by Лунтя on 06.06.2018.
 */
/*  Фрагмент списка списков */
public class ListsListFragment extends Fragment implements View.OnClickListener, IFilterContract {

    //private Button goToDetailBtn;
    private RecyclerView recyclerView;
    private TextView emptyList;
    private DataBase dbList;
    private FloatingActionButton fabAdd;
    IListsListActivityContract iListsListActivityContract;
    SwipeController swipeController = null;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_lists_list, container, false);

        fabAdd = rootView.findViewById(R.id.fabAddList);
        fabAdd.setOnClickListener(this);
        emptyList = rootView.findViewById(R.id.emptyList);

        recyclerView = rootView.findViewById(R.id.rvListsList);


        dbList = Application.Companion.getDB();

        refresh(null);

        return rootView;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof IListsListActivityContract) {
            iListsListActivityContract = (IListsListActivityContract) context;
        }
    }

    @Override
    public void onDetach() {
        iListsListActivityContract = null;
        super.onDetach();
    }

    private void refresh(@Nullable ArrayList<List> newLists) {

        if (newLists == null) {

            SQLiteDatabase db = this.dbList.getWritableDatabase();
            ArrayList<List> lists = new ArrayList<>();

            Cursor c = db.query(Constants.TABLE_NAME_LISTS_LIST, null, null,
                    null, null, null,
                    Constants.TABLE_ID + " asc");
            if (c.moveToFirst()) {

                // определяем номера столбцов по имени в выборке
                int idColIndex = c.getColumnIndex(Constants.TABLE_ID);
                int nameColIndex = c.getColumnIndex(Constants.TABLE_ITEM_NAME);
                do {
                    lists.add(new List(c.getInt(idColIndex), c.getString(nameColIndex)));
                    // переход на следующую строку
                    // а если следующей нет (текущая - последняя), то false - выходим из цикла
                } while (c.moveToNext());
            } else {
                c.close();
            }
            if (lists.size() == 0){
                emptyList.setVisibility(View.VISIBLE);
            }else{
                emptyList.setVisibility(View.GONE);
            }
            displayRW(lists);
        } else {
            displayRW(newLists);
        }
    }

    public void displayRW(ArrayList<List> lists) {

        AdapterListsList adapter = new AdapterListsList(lists, getActivity());
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
        RecyclerView.ItemAnimator itemAnimator = new DefaultItemAnimator();

        recyclerView.setAdapter(adapter);

        recyclerView.addItemDecoration(new RecyclerView.ItemDecoration() {
            @Override
            public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
                swipeController.onDraw(c);
            }
        });

        swipeController = new SwipeController(new SwipeControllerActions() {
            @Override
            public void onRightClicked(int position) {
                CRUDdb.deleteItemLists(lists.get(position));
                lists.remove(position);
                refresh(lists);
            }

            @Override
            public void onLeftClicked(int position) {
                super.onLeftClicked(position);
            }
        });

        ItemTouchHelper itemTouchhelper = new ItemTouchHelper(swipeController);
        itemTouchhelper.attachToRecyclerView(recyclerView);



        //recyclerView.addItemDecoration(new DividerItemDecoration(this.getContext(), DividerItemDecoration.VERTICAL));

        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL); // необязательно
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setItemAnimator(itemAnimator);

        //RecyclerTouchListener touchListener = new RecyclerTouchListener(this.getActivity(), recyclerView);
        /*touchListener.setClickable(new RecyclerTouchListener.OnRowClickListener() {
            @Override
            public void onRowClicked(int position) {
                Toast.makeText(getContext(),"onRowClicked", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onIndependentViewClicked(int independentViewID, int position) {
                Toast.makeText(getContext(),"onIndependentViewClicked", Toast.LENGTH_SHORT).show();
            }
        });
                .setSwipeOptionViews(R.id.delete_task, R.id.edit_task);*/

    }

    @Override
    public void onClick(View v) {

        switch (v.getId()) {
            case (R.id.fabAddList):
                iListsListActivityContract.onListListFragmentButtonClick();
                break;
        }

    }

    @Override
    public void onFilterMake(String filter) {
        setFilter(filter);
    }

    private void setFilter(String filter) {
        if (filter.length() == 0) {
            refresh(null);
        } else {
            SQLiteDatabase db = dbList.getWritableDatabase();
            ArrayList<List> lists = new ArrayList<>();

            Cursor c = db.rawQuery("select * from " + Constants.TABLE_NAME_LISTS_LIST +
                            " where " + Constants.TABLE_ITEM_NAME + " like '%"
                            + filter + "%'",
                    null);
            if (c.moveToFirst()) {

                // определяем номера столбцов по имени в выборке
                int idColIndex = c.getColumnIndex(Constants.TABLE_ID);
                int nameColIndex = c.getColumnIndex(Constants.TABLE_ITEM_NAME);
                do {
                    lists.add(new List(c.getInt(idColIndex), c.getString(nameColIndex)));
                    // переход на следующую строку
                    // а если следующей нет (текущая - последняя), то false - выходим из цикла
                } while (c.moveToNext());


            }
            c.close();
            refresh(lists);
        }
    }
}

File: app/src/main/java/net/sytes/kai_soft/letsbuyka/Lists/ListsListFragment.kt
package net.sytes.kai_soft.letsbuyka.Lists

import android.app.AlertDialog
import android.content.Context
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import net.sytes.kai_soft.letsbuyka.CRUDdb
import net.sytes.kai_soft.letsbuyka.IListActivityContract
import net.sytes.kai_soft.letsbuyka.SwipeController
import net.sytes.kai_soft.letsbuyka.IFilterContract
import net.sytes.kai_soft.letsbuyka.R
import net.sytes.kai_soft.letsbuyka.SwipeControllerActions

import java.util.ArrayList

/* Lists fragment   */
class ListsListFragment: Fragment(), View.OnClickListener, IFilterContract {

    companion object {
        const val myTag = "letsbuy_listsFragment"
    }

    private lateinit var recyclerView: RecyclerView
    private lateinit var emptyList: TextView
    private lateinit var fabAdd: FloatingActionButton
    private lateinit var iListActivityContract: IListActivityContract
    private lateinit var actualList: ArrayList<List>
    private lateinit var swipeController: SwipeController
    private lateinit var adapter: AdapterListsList

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Log.i(myTag, "onCreateView()")
        if (inflater != null){
            val rootView = inflater.inflate(R.layout.fragment_lists_list, container, false)
            fabAdd = rootView.findViewById(R.id.fabAddList)
            fabAdd.setOnClickListener(this)

            emptyList = rootView.findViewById(R.id.emptyList)

            recyclerView = rootView.findViewById(R.id.rvListsList)

            refresh()

            return rootView
        }
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onAttach(context: Context?) {
        Log.i(myTag, "onAttach()")
        if (context is IListActivityContract){
            iListActivityContract = context
        }
        super.onAttach(context)
    }

    /*  It gets list of lists and invokes displayRV method  */
    private fun refresh(newLists: ArrayList<List>? = null){
        if (newLists == null){
            Log.i(myTag, "refresh(null)")
        } else {
            Log.i(myTag, "refresh(newLists)")
        }

        if (newLists == null){
             actualList = CRUDdb.readFromTableLists()

            if (actualList.size == 0){
                emptyList.visibility = View.VISIBLE
            } else {
                emptyList.visibility = View.GONE
            }
            displayRV(actualList)
        } else {
            displayRV(newLists)
        }
    }

    /*  It shows recyclerView with data which refresh method found  */
    private fun displayRV(lists: ArrayList<List>){
        Log.i(myTag, "displayRV()")
        adapter = AdapterListsList(lists, activity)
        val linearLayoutManager = LinearLayoutManager(activity)
        val itemAnimator = DefaultItemAnimator()

        recyclerView.adapter = adapter

        swipeController = SwipeController(object : SwipeControllerActions(){
            override fun onRightSwiped(position: Int) {
                Log.i(myTag, "onRightSwiped()")
                val builder = AlertDialog.Builder(context)
                builder.setTitle(getString(R.string.deleting))
                        .setMessage(getString(R.string.deleteList))
                        .setPositiveButton(R.string.yes) { _, _ ->
                            adapter.notifyItemRemoved(position)
                            adapter.notifyItemRangeChanged(position, adapter.itemCount)
                            CRUDdb.deleteItemLists(lists[position])
                            lists.removeAt(position)

                            if (lists.size == 0){
                                emptyList.visibility = View.VISIBLE
                            }else{
                                emptyList.visibility = View.GONE
                            }
                        }
                        .setNegativeButton(R.string.no){ _, _ ->
                            adapter.notifyDataSetChanged()
                        }

                val alert = builder.create()
                alert.show()
            }

            override fun onLeftSwiped(position: Int) {
                Log.i(myTag, "onLeftSwiped()")
                iListActivityContract.onListItemLongClick(lists[position])
            }
        }, context)

        val itemTouchHelper = ItemTouchHelper(swipeController)
        itemTouchHelper.attachToRecyclerView(recyclerView)

        linearLayoutManager.orientation = LinearLayoutManager.VERTICAL
        recyclerView.layoutManager = linearLayoutManager
        recyclerView.itemAnimator = itemAnimator
    }

    override fun onClick(v: View?) {
        if (v != null){
            when (v.id){
                R.id.fabAddList ->{
                    Log.i(myTag, "onClick(fabAddList)")
                    iListActivityContract.onListFragmentButtonClick()
                }
            }
        }
    }

    override fun onFilterMake(newFilter: String?) {
        Log.i(myTag, "onFilterMake($newFilter)")
        adapter.notifyDataSetChanged()
        if (newFilter != null){
            setFilter(newFilter)
        }

    }

    private fun setFilter(filter: String){
        Log.i(myTag, "onFilterMake($filter)")
        if (filter.isEmpty()){
            refresh()
        } else {
            val lists = CRUDdb.readFromTableLists(filter)
            refresh(lists)
        }
    }
}